alf/utils/dist_utils.py

```diff
@@ -58,6 +58,117 @@ class NewCls(cls):
+    and `tensorflow <https://github.com/tensorflow/probability/blob/v0.12.2/tensorflow_probability/python/bijectors/softplus.py#L61-L189>`_.
+    """
+    domain = constraints.real
+    codomain = constraints.positive
+    bijective = True
+    sign = +1
+
+    def __init__(self, hinge_softness=1.):
+        """
+        Args:
+            hinge_softness (float): this positive parameter changes the transition
+                slope. A higher softness results in a smoother transition from
+                0 to identity.
+        """
+        super().__init__(cache_size=1)
+        self._hinge_softness = float(hinge_softness)
+        assert self._hinge_softness > 0, "Must be a positive softness number!"
+
+    def __eq__(self, other):
+        return isinstance(other, Softplus)
+
+    def _call(self, x):
+        return (nn.functional.softplus(x / self._hinge_softness) *
+                self._hinge_softness)
+
+    def _inverse(self, y):
+        return (y / self._hinge_softness).expm1().log() * self._hinge_softness
+
+    def log_abs_det_jacobian(self, x, y):
+        return -nn.functional.softplus(-x / self._hinge_softness)
+
+
+@alf.configurable
+def Softlower(low, hinge_softness=1.):
+    """Create a Softlower transform by composing the Softplus and Affine
+    transforms. Mathematically, ``softlower(x, low) = softplus(x - low) + low``.
+
+    Args:
+        low (float|Tensor): the lower bound
+        hinge_softness (float): this positive parameter changes the transition
+                slope. A higher softness results in a smoother transition from
+                ``low`` to identity.
+    """
+    return td.transforms.ComposeTransform([
+        AffineTransform(loc=-low, scale=1.),
+        Softplus(hinge_softness=hinge_softness),
+        AffineTransform(loc=low, scale=1.)
+    ])
+
+
+@alf.configurable
+def Softupper(high, hinge_softness=1.):
+    """Create a Softupper transform by composing the Softplus and Affine
+    transforms. Mathematically, ``softupper(x, high) = -softplus(high - x) + high``.
+
+    Args:
+        high (float|Tensor): the upper bound
+        hinge_softness (float): this positive parameter changes the transition
+                slope. A higher softness results in a smoother transition from
+                identity to ``high``.
+    """
+    return td.transforms.ComposeTransform([
+        AffineTransform(loc=high, scale=-1.),
+        Softplus(hinge_softness=hinge_softness),
+        AffineTransform(loc=high, scale=-1.)
+    ])
+
+
+@alf.configurable
+def Softclip(low, high, hinge_softness=1.):
+    """Create a Softclip transform by composing Softlower, Softupper, and Affine
+    transforms, adapted from `tensorflow <https://www.tensorflow.org/probability/api_docs/python/tfp/bijectors/SoftClip>`_.
+    Mathematically,
+
+    .. code-block:: python
+
+        clipped = softupper(softlower(x, low), high)
+        softclip(x) = (clipped - high) / (high - softupper(low, high)) * (high - low) + high
```

Comment:
This one is indeed better. But it seems that I can't use ComposedTransform to configure this transform from existing ones (there is no built-in transform that combines two transforms). Do you think we should write a custom Softclip transform from scratch (like Softplus)?

Reply:
> It seems that upper and lower are not treated symmetrically, which feels not very nice.
> The following is a symmetric one which has analytical inverse and bounded within (low, high)
> 
> ```
> softlow(x, low) + softhigh(x, high) - x
> ```

This one is indeed better. But it seems that I can't use ComposedTransform to configure this transform from existing ones (there is no built-in transform that combines two transforms). Do you think we should write a custom Softclip transform from scratch (like Softplus)?

Reply:
It's not that difficult to implement it directly. The actual formula is quite simple if you plugin softlow and softhigh.

Reply:
> It's not that difficult to implement it directly. The actual formula is quite simple if you plugin softlow and softhigh.

I think the main difficulty is to derive the inverse function and log_abs_jacobian and verify their correctness.

Reply:
Inverse is not that difficult for this one, not much difficult to calculate compared to softhigh. Jacobian is just the absolute value of first order derivative.

Reply:
It seems that this solution has some issues when x is very large or small. The clipped values will be slightly beyond the (low, high) range, while the TFP solution won't have this issue. 

```python
import torch
import alf

def softclip(x, low, high, hinge_softness=1.):
    """Softly bound ``x`` in between ``[low, high]``, namely,
    ``softlower(x, low) + softupper(x, high) - x``. Unlike ``softclip_tf``,
    this transform is symmetric regarding the lower and upper bound when
    squashing.

    Args:
        x (Tensor): input
        low (float|Tensor): the lower bound
        high (float|Tensor): the upper bound
        hinge_softness (float): this positive parameter changes the transition
            slope. A higher softness results in a smoother transition from
            ``low`` to ``high``. Default to 1.
    """
    return (alf.math.softlower(x, low, hinge_softness)
            + alf.math.softupper(x, high, hinge_softness) - x)

if __name__ == "__main__":
    torch.set_printoptions(precision=8)
    x = torch.arange(-5, 5, 0.01)
    b = 1
    hinge_softness = 1e-1
    clipped_x = softclip(x, -b, b, hinge_softness)
    print(clipped_x)
    print(clipped_x <= b)
```

Reply:
Interesting. Mathematically, it is guaranteed to be within (l, h) and increasing. Not sure if there is a way to write it so that it has better numerical precision.

Reply:
Sorry, indeed it is not monotonically increasing close to the end

Reply:
Sorry again, it is actually monotonically increasing

Reply:
Colab here: https://colab.research.google.com/drive/1677E5AePlK5fx2uSWS3p84BEjJ5P3Qel?usp=sharing

Reply:
added a new softclip function